src/main/java/duke/Parser.java

```diff
@@ -0,0 +1,132 @@
+            throws IncompleteInputException,
+            WrongMarkException,
+            InvalidInputException {
+        commandGiven = commandGiven.trim();
+        String[] splittedCommands = commandGiven.split(" ");
+        String commandType = splittedCommands[0];
+        String command = "";
+        switch (commandType) {
+        case "list":
+            ui.separatorLines();
+            command = tasks.printTasks();
+            break;
+        case "mark":
+            try {
+                Task task = tasks.get(Integer.parseInt(splittedCommands[1]) - 1);
+                if (Objects.equals(task.getStatusIcon(), "X")) {
+                    command = ui.showAlreadyDone();
+                } else {
+                    ui.separatorLines();
+                    command = task.setAsDone(task);
+                }
+                storage.save(tasks);
+            } catch (IndexOutOfBoundsException | NullPointerException e) {
+                command = ui.showInvalidIndex();
+            } catch (IOException e) {
+                command = String.valueOf(e);
+                throw new RuntimeException(e);
+            }
+            break;
+        case "unmark":
+            try {
+                Task task = tasks.get(Integer.parseInt(splittedCommands[1]) - 1);
+                if (Objects.equals(task.getStatusIcon(), " ")) {
+                    command = ui.showAlreadyUndone();
+                } else {
+                    ui.separatorLines();
+                    command = task.setAsUndone(task);
+                }
+                storage.save(tasks);
+            } catch (IndexOutOfBoundsException | NullPointerException e) {
+                command = ui.showInvalidIndex();
+            } catch (IOException e) {
+                command = String.valueOf(e);
+                throw new RuntimeException(e);
+            }
+            break;
+        case "delete":
+            try {
+                Task task = tasks.get(Integer.parseInt(splittedCommands[1]) - 1);
+                tasks.deleteTask(Integer.parseInt(splittedCommands[1]));
+                command = ui.deleteTaskMessage(task);
+            } catch (IndexOutOfBoundsException | NullPointerException e) {
+                command = ui.showInvalidIndex();
+            }
+            break;
+        case "find":
+            ui.separatorLines();
+            try {
+                String keyword = splittedCommands[1];
+                command = tasks.findTasks(keyword);
+            } catch (IndexOutOfBoundsException | NullPointerException e) {
+                command = "find what eh! ";
+                throw new IncompleteInputException("find what eh! ");
+            }
+            break;
+        case "bye":
+            try {
+                storage.save(tasks);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+            command = "bye";
+            break;
+        default:
```

Comment:
Why do you perform checks here instead of just directly giving back an error warning message that such an action doesn't exist

Reply:
It is because the error can be either Input or Incomplete error so it would be better to check here